### ISSUE TITLE
Rustbuild rebuilt when checkout branch and checkout back

### ISSUE THREAD

**Issue by lzutao** [rust-lang/rust#62648]
For example, when I am on my branch `feature`, I checkout to `master` and check back to `feature`, 
rustbuid rebuilt from stage 0.

**Comment by Mark-Simulacrum** [comment]
Can you run `git log -p` in both master and your feature branch and gist the output?

**Comment by lzutao** [comment]
Could you give more specific command? The output of `git log -p` is too big and slow.

**Comment by Mark-Simulacrum** [comment]
`git log -p -- src/bootstrap master...feature`, perhaps. The (primary) goal here is to determine whether anything in bootstrap is changing in that commit range; if that's already been ascertained, then running with `CARGO_LOG=cargo::core::compiler::fingerprint=info` may yield something helpful, too.

**Comment by lzutao** [comment]
<details>
	<summary>% git log -p deprecated-try-macro..tidy-lldb deprecated-try-macro</summary>

```diff
commit d649ff3c0c048520c054a539d7c9eb9cf77b127a (origin/tidy-lldb, tidy-lldb)
Author: Lzu Tao <taolzu@gmail.com>
Date:   Sun Jul 14 04:19:39 2019 +0000

    Check for lldb existences

diff --git a/src/etc/rust-lldb b/src/etc/rust-lldb
index 0eb99423df..7b9b40e6b4 100755
--- a/src/etc/rust-lldb
+++ b/src/etc/rust-lldb
@@ -3,25 +3,30 @@
 # Exit if anything fails
 set -e
 
-# Find out where to look for the pretty printer Python module
-RUSTC_SYSROOT=`rustc --print sysroot`
-
 # Find the host triple so we can find lldb in rustlib.
-host=`rustc -vV | sed -n -e 's/^host: //p'`
+host=$(rustc -vV | sed -n -e 's/^host: //p')
+
+# Find out where to look for the pretty printer Python module
+RUSTC_SYSROOT=$(rustc --print sysroot)
+RUST_LLDB="$RUSTC_SYSROOT/lib/rustlib/$host/bin/lldb"
 
 lldb=lldb
-if [ -f "$RUSTC_SYSROOT/lib/rustlib/$host/bin/lldb" ]; then
-    lldb="$RUSTC_SYSROOT/lib/rustlib/$host/bin/lldb"
+if [ -f "$RUST_LLDB" ]; then
+    lldb="$RUST_LLDB"
 else
-    LLDB_VERSION=`"$lldb" --version 2>/dev/null | head -1 | cut -d. -f1`
+    if ! command -v "$lldb" > /dev/null; then
+        echo "$lldb not found! Please install it." >&2
+        exit 1
+    else
+        LLDB_VERSION=$("$lldb" --version | cut -d ' ' -f3)
 
-    if [ "$LLDB_VERSION" = "lldb-350" ]
-    then
-        echo "***"
-       echo \
-"WARNING: This version of LLDB has known issues with Rust and cannot \
-display the contents of local variables!"
-        echo "***"
+        if [ "$LLDB_VERSION" = "3.5.0" ]; then
+            cat << EOF >&2
+***
+WARNING: This version of LLDB has known issues with Rust and cannot display the contents of local variables!
+***
+EOF
+        fi
     fi
 fi
 
```

</details>

**Comment by Mark-Simulacrum** [comment]
What about the `CARGO_LOG` run? Do you expect that the branches have different src/bootstrap or src/stage0.txt contents?

**Comment by lzutao** [comment]
The gist is here: https://gist.github.com/lzutao/7b6bd404b75226452fffb65d298bb046

> Do you expect that the branches have different src/bootstrap or src/stage0.txt contents?

No, I don't.

**Comment by Mark-Simulacrum** [comment]
*rustbuild* isn't being rebuilt here; I think I misinterpreted your earlier statement. Can you show the diff between the two branches? It looks like (presumably) `std` has changed at least...

**Comment by lzutao** [comment]
The diff is here: https://gist.github.com/lzutao/3dfed8c8bd1815e70949c05e50d1cefa

The problem is: I traveled between branches and went back to the one that have been 
built sucessully. rustbuild build that branch from std stage0.

**Comment by Mark-Simulacrum** [comment]
This diff changes libcore, which means that switching between these two branches will rebuild. Cargo uses timestamps to determine if a rebuild is needed, so it's unsurprising to me that this is happening for you. This is most likely not a rustbuild bug, and is probably intentional behavior.

**Comment by lzutao** [comment]
Yes, the diff changed libcore. Even though I went back to the branch that I built earlier without doing
any changes. 

I have just done some researches. Like you said, when switching between branches git will update
the timestamp for any changed files between those two branches. Perhaps I compared cargo 
with mesonbuild. Mesonbuild uses hashes to determine files changed.

I would close this issue as not a bug then.
